MelonLoader/Melons/MelonBase.cs

```diff
@@ -578,61 +592,97 @@ public virtual bool Register()
                 LoggerInstance = new MelonLogger.Instance(string.IsNullOrEmpty(ID) ? Info.Name : $"{ID}:{Info.Name}", ConsoleColor);
             HarmonyInstance = new HarmonyLib.Harmony($"{Assembly.FullName}:{Info.Name}");
 
+            Registered = true; // this has to be true before the melon can subscribe to any events
+            RegisterCallbacks();
+
             try
             {
-                InitializeMelon();
+                OnInitializeMelon();
             }
             catch (Exception ex)
             {
                 MelonLogger.Error($"Failed to register {MelonTypeName} '{Location}': Melon failed to initialize!");
                 MelonLogger.Error(ex.ToString());
+                Registered = false;
                 return false;
             }
 
+            if (!RegisterInternal())
+                return false;
+
             _registeredMelons.Add(this);
-            Registered = true;
 
             if (!HarmonyDontPatchAll)
                 HarmonyInstance.PatchAll(Assembly);
             RegisterTypeInIl2Cpp.RegisterAssembly(Assembly);
 
+
             PrintLoadInfo();
-            onMelonRegistered?.Invoke(this);
+
+            OnRegister.Invoke();
+            OnMelonRegistered.Invoke(this);
             return true;
         }
 
+        protected internal virtual bool RegisterInternal() => true;
+        protected internal virtual bool UnregisterInternal() => true;
+
+        protected internal virtual void RegisterCallbacks()
+        {
+            MelonEvents.OnPreInitialization.Subscribe(OnPreSupportModule);
+            MelonEvents.OnApplicationStart.Subscribe(OnApplicationStart);
+            MelonEvents.OnApplicationLateStart.Subscribe(OnApplicationLateStart);
+            MelonEvents.OnApplicationQuit.Subscribe(OnApplicationQuit);
+            MelonEvents.OnUpdate.Subscribe(OnUpdate);
+            MelonEvents.OnLateUpdate.Subscribe(OnLateUpdate);
+            MelonEvents.OnGUI.Subscribe(OnGUI);
+            MelonEvents.OnFixedUpdate.Subscribe(OnFixedUpdate);
+
+            MelonPreferences.OnPreferencesLoaded.Subscribe(OnPreferencesLoaded);
+            MelonPreferences.OnPreferencesLoaded.Subscribe((x) => OnPreferencesLoaded()); // No params sig
+            MelonPreferences.OnPreferencesSaved.Subscribe(OnPreferencesSaved);
+            MelonPreferences.OnPreferencesSaved.Subscribe((x) => OnPreferencesSaved()); // No params sig
+            MelonPreferences.OnPreferencesSaved.Subscribe((x) => OnModSettingsApplied());
+        }
+
         /// <summary>
-        /// If the Name is registered, returns the registered Melon. Otherwise, returns <see langword="null"/>.
+        /// Tries to find a register Melon to matches the given Info.
```

Comment:
to matches -> which matches

Reply:
bro i actually have no damn idea how i messed this comment up this bad lmao. It must have been like 3am

Reply:
https://github.com/SlidyDev/MelonLoader/commit/ed9605a47d7518b65b90fe9b6c8a3cdd7ce8ff85